feat: add `middleware` option to Parse Server config

okobsamoht opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [ ] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

the 'middleware' option only works if parse-server is used as command line binary.
when using parse-server as an express middleware the option 'middleware' have no effect.


### Approach
<!-- Add a description of the approach in this PR. -->

i uses the same condition of the middleware in ParseServer.start({...options}, callback) to let parse-server handle the middleware in ParseServer.app({...options}) 

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)

mtrezza committed:
ci: fix node engine check (#7891)

dblythy committed:
fix: return correct response when revert is used in beforeSave (#7839)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.1 [skip ci]

## [5.2.1-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.0...5.2.1-alpha.1) (2022-03-26)

### Bug Fixes

* return correct response when revert is used in beforeSave ([#7839](https://github.com/parse-community/parse-server/issues/7839)) ([f63fb2b](https://github.com/parse-community/parse-server/commit/f63fb2b338c908f0e7a648d338c26b9daa50c8f2))

dblythy committed:
perf: reduce database operations when using the constant parameter in Cloud Function validation (#7892)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.2 [skip ci]

## [5.2.1-alpha.2](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.1...5.2.1-alpha.2) (2022-03-26)

### Performance Improvements

* reduce database operations when using the constant parameter in Cloud Function validation ([#7892](https://github.com/parse-community/parse-server/issues/7892)) ([48bd512](https://github.com/parse-community/parse-server/commit/48bd512eeb47666967dff8c5e723ddc5b7801daa))

github-actions[bot] committed:
feat: add MongoDB 5.1 compatibility (#7682)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.1 [skip ci]

# [5.3.0-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.2...5.3.0-alpha.1) (2022-03-27)

### Features

* add MongoDB 5.1 compatibility ([#7682](https://github.com/parse-community/parse-server/issues/7682)) ([90155cf](https://github.com/parse-community/parse-server/commit/90155cf1680e5e0499b0000e071c6cb0ce3aef96))

mtrezza committed:
fix: security upgrade parse push adapter from 4.1.0 to 4.1.2 (#7893)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.2 [skip ci]

# [5.3.0-alpha.2](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.1...5.3.0-alpha.2) (2022-03-27)

### Bug Fixes

* security upgrade parse push adapter from 4.1.0 to 4.1.2 ([#7893](https://github.com/parse-community/parse-server/issues/7893)) ([ef56e98](https://github.com/parse-community/parse-server/commit/ef56e98ef65041b4d3b7b82cce3473269c27f6fd))

mtrezza committed:
feat: add MongoDB 5.2 support (#7894)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.3 [skip ci]

# [5.3.0-alpha.3](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.2...5.3.0-alpha.3) (2022-03-27)

### Features

* add MongoDB 5.2 support ([#7894](https://github.com/parse-community/parse-server/issues/7894)) ([6b4b358](https://github.com/parse-community/parse-server/commit/6b4b358f0842ae920e45652f5e8b2afebc6caf3a))

snyk-bot committed:
refactor: upgrade mongodb from 4.3.1 to 4.4.1 (#7906)

github-actions[bot] committed:
ci: bump environment (#7907)

Moumouls committed:
refactor: add missing schema definitions (#7917)

dblythy committed:
fix: custom database options are not passed to MongoDB GridFS (#7911)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.4 [skip ci]

# [5.3.0-alpha.4](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.3...5.3.0-alpha.4) (2022-04-04)

### Bug Fixes

* custom database options are not passed to MongoDB GridFS ([#7911](https://github.com/parse-community/parse-server/issues/7911)) ([a72b384](https://github.com/parse-community/parse-server/commit/a72b384f76137a3d83ffb69f65cb25aff1bbab4f))

snyk-bot committed:
refactor: upgrade @apollo/client from 3.5.8 to 3.5.9 (#7921)

dependabot[bot] committed:
fix: security upgrade moment from 2.29.1 to 2.29.2 (#7931)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.5 [skip ci]

# [5.3.0-alpha.5](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.4...5.3.0-alpha.5) (2022-04-09)

### Bug Fixes

* security upgrade moment from 2.29.1 to 2.29.2 ([#7931](https://github.com/parse-community/parse-server/issues/7931)) ([6b68593](https://github.com/parse-community/parse-server/commit/6b68593eaec17e8b183899d2b92699c9ede7625b))

marvinroger committed:
fix: peer dependency mismatch for GraphQL dependencies (#7934)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.6 [skip ci]

# [5.3.0-alpha.6](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.5...5.3.0-alpha.6) (2022-04-11)

### Bug Fixes

* peer dependency mismatch for GraphQL dependencies ([#7934](https://github.com/parse-community/parse-server/issues/7934)) ([b7a1d76](https://github.com/parse-community/parse-server/commit/b7a1d7617b4bcac677cecedfeb6ac4a27447083b))

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.2 to 8.2.4 (#7932)

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.4 to 8.2.6 (#7935)

okobsamoht committed:
feat: add support of middlewares when using parse-server as express middleware

okobsamoht committed:
feat: add support of middlewares when using parse-server as express middleware

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.6 to 8.2.7 (#7941)

mtrezza commented on src/ParseServer.js:
```diff
@@ -184,6 +184,19 @@ class ParseServer {
     api.use(middlewares.allowMethodOverride);
     api.use(middlewares.handleParseHeaders);
 
+    // add support of middlewares
```
Remove comments

okobsamoht replied:
hi
* its not a new config option

mtrezza commented on src/ParseServer.js:
```diff
@@ -184,6 +184,19 @@ class ParseServer {
     api.use(middlewares.handleParseHeaders);
 
+    // add support of middlewares
+    if (options.middleware) {
```
See [contribution guide](https://github.com/parse-community/parse-server/blob/alpha/CONTRIBUTING.md#parse-server-configuration) for how to add a new config option to Parse Server; the option needs to be validated, etc.

okobsamoht replied:
hi.
* the comment is removed
* i do not add a new config option. the "middleware" option already exists

okobsamoht committed:
feat: add support of middlewares when using parse-server as express middleware - remove comment

okobsamoht commented:
for the middleware config option you can see in:
`https://github.com/parse-community/parse-server/blob/alpha/src/Options/Definitions.js`
it already exist

mtrezza commented:
Right, it already exists, there is even a test case:

https://github.com/parse-community/parse-server/blob/4c29d4d23b67e4abaf25803fe71cae47ce1b5957/spec/index.spec.js#L489-L508

If the test already passes, why is this PR needed?

okobsamoht commented:
it works only when i run parse-server like this:
`parse-server --appId APPLICATION_ID --masterKey MASTER_KEY --databaseURI mongodb://localhost/test --middleware [middleware_path] `

but it does not work when i do this:
```var express = require('express');
var ParseServer = require('parse-server').ParseServer;

var app = express();
var api = new ParseServer({ 
   ...options,
   middleware: 'middleware'
});```

app.use('/parse', api);

var port = 1337;
app.listen(port, function() {
  console.log('parse-server-example running on port ' + port + '.');
});
```
i was struggling on it

mtrezza commented:
What's the difference between your example and the test?

okobsamoht commented:
the diffrerence between the two is explained here:
* for using parseserver's internal express app:
https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/src/ParseServer.js#L252
* for starting parseserver inside another express app:
https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/src/ParseServer.js#L152

okobsamoht commented:
the tests calls the parseserver's internal express app :
https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/spec/helper.js#L169

the lines of code handling the 'middleware' option is inside that function

so i copy the logic from the ParseServer.start() function:
https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/src/ParseServer.js#L257

https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/src/ParseServer.js#L259

to the ParseServer.app() function:
https://github.com/parse-community/parse-server/blob/8b919613f0babe1bbf524fb391b0d3853e029e68/src/ParseServer.js#L154

dplewis committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#7944)

mtrezza committed:
Merge branch 'alpha' into alpha-middleware-fix

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.7 to 8.2.8 (#7945)

mtrezza committed:
Merge branch 'alpha' into alpha-middleware-fix

snyk-bot committed:
fix: security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 (#7948)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.7 [skip ci]

# [5.3.0-alpha.7](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.6...5.3.0-alpha.7) (2022-04-25)

### Bug Fixes

* security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 ([#7948](https://github.com/parse-community/parse-server/issues/7948)) ([20fc4e2](https://github.com/parse-community/parse-server/commit/20fc4e23b53c91aac657f894bd70d049b7525c37))

okobsamoht committed:
Merge branch 'parse-community:alpha' into alpha-middleware-fix

mtrezza committed:
refactor: upgrade winston from 3.5.1 to 3.7.2 (#7954)

dblythy commented:
I can confirm that I have written a test that fails on `alpha` but passes on this branch. I have provided it in my review.

Could you please also link the related issue and amend the PR to follow the template. This process ensures that future users experiencing similar issues can easily identify the problem and related fix. 😊

dblythy commented on src/ParseServer.js:
```diff
@@ -184,6 +184,18 @@ class ParseServer {
+      }
+      api.use(middleware);
+    }
+
```
```js
it('can create server with middleware', async () => {
    const obj = {
      middleware: function (req, res, next) {
        next();
      },
    };
    const spy = spyOn(obj, 'middleware').and.callThrough();
    const parseServer = await new Promise((resolve, reject) => {
      const server = ParseServer.ParseServer({
        ...defaultConfiguration,
        ...{
          appId: 'anOtherTestApp',
          masterKey: 'anOtherTestMasterKey',
          serverURL: 'http://localhost:12667/parse',
          serverStartComplete(error) {
            if (error) {
              reject(error);
            } else {
              resolve(server);
            }
          },
          middleware: obj.middleware,
        },
      });
    });
    const app = express();
    app.use('/parse', parseServer);
    await new Promise(resolve => app.listen(12667, resolve));
    expect(Parse.applicationId).toEqual('anOtherTestApp');
    expect(Parse.serverURL).toEqual('http://localhost:12667/parse');
    await new Parse.Query('AnObject').find();
    expect(spy).toHaveBeenCalled();
  });
```

Can you please add this to [index.spec.js](https://github.com/parse-community/parse-server/blob/alpha/spec/index.spec.js#L509)

okobsamoht replied:
done!
thanks for the review.

dblythy replied:
sorry, you'll also need to close the server after the test finishes

replace

`await new Promise(resolve => app.listen(12667, resolve));` with `const httpServer = app.listen(12667)`

then call `httpServer.close()` at the end of the test

dblythy commented on src/ParseServer.js:
```diff
@@ -184,6 +184,18 @@ class ParseServer {
+      if (typeof options.middleware == 'string') {
+        middleware = require(path.resolve(process.cwd(), options.middleware));
+      } else if (typeof options.middleware == 'function') {
+        middleware = options.middleware; // use as-is let express fail
```
No need for inline comment

okobsamoht replied:
done!

dblythy commented on src/ParseServer.js:
```diff
@@ -184,6 +184,18 @@ class ParseServer {
+      } else if (typeof options.middleware == 'function') {
+        middleware = options.middleware; // use as-is let express fail
+      } else {
+        throw "argument 'middleware' must either be a string or a function";
```
might need to add coverage for this:

```js
it('cannot create middleware with invalid type', async () => {
    await expectAsync(new Promise((resolve, reject) => {
      const server = ParseServer.ParseServer({
        ...defaultConfiguration,
        ...{
          serverStartComplete(error) {
            if (error) {
              reject(error);
            } else {
              resolve(server);
            }
          },
          middleware: true
        },
      });
    })).toBeRejectedWith("argument 'middleware' must either be a string or a function");
  });
```

okobsamoht replied:
done!

dblythy reviewed (changes requested):
Looks good, let's just add some test cases to be sure.

mtrezza committed:
Merge branch 'alpha' into alpha-middleware-fix

okobsamoht committed:
SPEC: can create server with middleware

okobsamoht committed:
SPEC: cannot create middleware with invalid type

okobsamoht committed:
removed inline comment

Commit:
Merge remote-tracking branch 'source/alpha' into alpha

# Conflicts:
#	package-lock.json
#	package.json

Commit:
Merge remote-tracking branch 'origin/alpha' into alpha-middleware-fix

dblythy commented on spec/index.spec.js:
```diff
@@ -531,6 +531,58 @@ describe('server', () => {
+    });
+    const app = express();
+    app.use('/parse', parseServer);
+    await new Promise(resolve => app.listen(12667, resolve));
```
I think you need to close the server after the tests..

Can you refactor

```const httpServer = app.listen(12667, resolve);```

then add `httpServer.close()` to the end of the test

okobsamoht replied:
done.

okobsamoht committed:
Merge branch 'parse-community:alpha' into alpha-middleware-fix

okobsamoht committed:
refactore and close server after middleware test

okobsamoht committed:
Merge branch 'alpha' into alpha-middleware-fix

dblythy commented:
Looks mostly good - might just have to fix lint and tests

okobsamoht committed:
Merge branch 'parse-community:alpha' into alpha-middleware-fix

mtrezza committed:
Merge branch 'alpha' into alpha-middleware-fix

mtrezza commented:
@okobsamoht could you take a look at this PR to get this ready for merging?

okobsamoht committed:
Merge branch 'parse-community:alpha' into alpha-middleware-fix

okobsamoht reviewed (commented):
reviewed

okobsamoht committed:
fix 'resolve' is not defined

mtrezza committed:
Merge branch 'alpha' into alpha-middleware-fix

dblythy commented:
Come to think of it do you really need `config.middleware` when using the Parse Server express constructor? Could you just use:

```js

const app = express();
app.use(…custom middleware here)
app.use("/parse", new ParseServer(config));
app.listen(1337)
```

mtrezza reviewed (commented):
Waiting for reply to https://github.com/parse-community/parse-server/pull/7942#issuecomment-1172962526

okobsamoht committed:
Merge branch 'parse-community:alpha' into alpha-middleware-fix

okobsamoht commented:
> Come to think of it do you really need `config.middleware` when using the Parse Server express constructor? Could you just use:
> 
> ```js
> const app = express();
> app.use(…custom middleware here)
> app.use("/parse", new ParseServer(config));
> app.listen(1337)
> ```

Hi @dblythy 
when i do that, my custom middleware have no access to ParseServer runtime stuffs like Controllers, Adapters, Options.
and when i log `req.config` i always get `undefined` in my custom middleare.

okobsamoht commented:
i make this quick setup:
```js
var express = require('express');
var ParseServer = require('parse-server').ParseServer;

var api = new ParseServer({
    appId: "APP_ID",
    appNAme: "APP_NAME",
    javascriptKey: "JAVASCRIPT_KEY",
    masterKey: "MASTER_KEY",
    directAccess: true,
    enforcePrivateUsers: true,
    port: 3000,
    mountPath: '/api',
    cloud:'cloudCode',
    middleware:(req, res, next)=>{
        console.log('parse middleware_________________________________________________________________________________')
        console.log(req.config)
        next()
    }
});

var app = express();

app.use((req, res, next)=>{
    console.log('express middleware___________________________________________________________________________________')
    console.log(req.config)
    next()
})
// parse server
app.use('/api', api);

var httpServer = require('http').createServer(app);
httpServer.listen(3000);

```

and each time a ran a request here is the result:

```log
express middleware___________________________________________________________________________________
undefined
parse middleware_________________________________________________________________________________
<ref *2> Config {
  applicationId: 'APP_ID',
  appId: 'APP_ID',
  appNAme: 'APP_NAME',
  javascriptKey: 'JAVASCRIPT_KEY',
  masterKey: 'MASTER_KEY',
  directAccess: true,
  enforcePrivateUsers: true,
  port: 3000,
  mountPath: '/api',
  cloud: 'cloudCode',
  middleware: [Function: middleware],
  allowClientClassCreation: true,
  allowCustomObjectId: false,
  cacheMaxSize: 10000,
  cacheTTL: 5000,
  collectionPrefix: '',
  customPages: {},
  databaseURI: 'mongodb://localhost:27017/parse',
  emailVerifyTokenReuseIfValid: false,
  enableAnonymousUsers: true,
  enableExpressErrorHandler: false,
  expireInactiveSessions: true,
  fileUpload: {
    enableForAnonymousUser: false,
    enableForPublic: false,
    enableForAuthenticatedUser: true
  },
  graphQLPath: '/graphql',
  host: '0.0.0.0',
  idempotencyOptions: { ttl: 300, paths: [] },
  logsFolder: './logs/',
  masterKeyIps: [],
  maxUploadSize: '20mb',
  mountGraphQL: false,
  mountPlayground: false,
  objectIdSize: 10,
  pages: {
    enableRouter: false,
    enableLocalization: false,
    localizationJsonPath: undefined,
    localizationFallbackLocale: 'en',
    placeholders: {},
    forceRedirect: false,
    pagesPath: './public',
    pagesEndpoint: 'apps',
    customUrls: {},
    customRoutes: []
  },
  playgroundPath: '/playground',
  preserveFileName: false,
  preventLoginWithUnverifiedEmail: false,
  protectedFields: { _User: { '*': [Array] } },
  requestKeywordDenylist: [
    { key: '_bsontype', value: 'Code' },
    { key: 'constructor' },
    { key: '__proto__' }
  ],
  revokeSessionOnPasswordReset: true,
  scheduledPush: false,
  security: { enableCheck: false, enableCheckLog: false },
  sessionLength: 31536000,
  verifyUserEmails: false,
  jsonLogs: false,
  verbose: false,
  level: undefined,
  serverURL: 'http://localhost:3000/api',
  loggerController: LoggerController {
    options: {
      jsonLogs: false,
      logsFolder: './logs/',
      verbose: false,
      logLevel: undefined,
      silent: undefined,
      maxLogFiles: undefined
    },
    appId: 'APP_ID',
    debug: [Function (anonymous)],
    verbose: [Function (anonymous)],
    silly: [Function (anonymous)],
    [Symbol()]: WinstonLoggerAdapter {}
  },
  filesController: FilesController {
    options: { preserveFileName: false },
    appId: 'APP_ID',
    [Symbol()]: GridFSBucketAdapter {
      _databaseURI: 'mongodb://localhost:27017/parse',
      _algorithm: 'aes-256-gcm',
      _encryptionKey: null,
      _mongoOptions: [Object]
    }
  },
  userController: UserController {
    options: { verifyUserEmails: false },
    appId: 'APP_ID',
    [Symbol()]: undefined
  },
  pushController: PushController {},
  hasPushScheduledSupport: false,
  hasPushSupport: false,
  pushWorker: PushWorker {
    adapter: ParsePushAdapter {
      supportsPushTracking: true,
      validPushTypes: [Array],
      senderMap: {},
      feature: [Object]
    },
    channel: 'APP_ID-parse-server-push',
    subscriber: Consumer {
      _events: [Object: null prototype],
      _eventsCount: 1,
      _maxListeners: undefined,
      emitter: [EventEmitter],
      [Symbol(kCapture)]: false
    }
  },
  pushControllerQueue: PushQueue {
    channel: 'APP_ID-parse-server-push',
    batchSize: 100,
    parsePublisher: Publisher { emitter: [EventEmitter] }
  },
  analyticsController: AnalyticsController {
    options: undefined,
    appId: undefined,
    [Symbol()]: AnalyticsAdapter {}
  },
  cacheController: <ref *1> CacheController {
    options: {},
    appId: 'APP_ID',
    role: SubCache { prefix: 'role', cache: [Circular *1], ttl: undefined },
    user: SubCache { prefix: 'user', cache: [Circular *1], ttl: undefined },
    graphQL: SubCache {
      prefix: 'graphQL',
      cache: [Circular *1],
      ttl: undefined
    },
    [Symbol()]: InMemoryCacheAdapter { cache: [LRUCache] }
  },
  parseGraphQLController: ParseGraphQLController {
    databaseController: DatabaseController {
      adapter: [MongoStorageAdapter],
      options: [Object],
      idempotencyOptions: [Object],
      _transactionalSession: null
    },
    cacheController: <ref *1> CacheController {
      options: {},
      appId: 'APP_ID',
      role: [SubCache],
      user: [SubCache],
      graphQL: [SubCache],
      [Symbol()]: [InMemoryCacheAdapter]
    },
    isMounted: false,
    configCacheKey: 'config'
  },
  liveQueryController: LiveQueryController {
    classNames: Set(0) {},
    liveQueryPublisher: ParseCloudCodePublisher { parsePublisher: [Publisher] }
  },
  database: DatabaseController {
    adapter: MongoStorageAdapter {
      _uri: 'mongodb://localhost:27017/parse',
      _collectionPrefix: '',
      _mongoOptions: [Object],
      _onchange: [Function (anonymous)],
      _maxTimeMS: undefined,
      canSortOnJoinTables: true,
      enableSchemaHooks: false,
      connectionPromise: [Promise],
      client: [MongoClient],
      database: [Db]
    },
    options: [Circular *2],
    idempotencyOptions: { ttl: 300, paths: [] },
    schemaPromise: null,
    _transactionalSession: null
  },
  hooksController: HooksController {
    _applicationId: 'APP_ID',
    _webhookKey: undefined,
    database: DatabaseController {
      adapter: [MongoStorageAdapter],
      options: [Object],
      idempotencyOptions: [Object],
      _transactionalSession: null
    }
  },
  authDataManager: {
    getValidatorForProvider: [Function: getValidatorForProvider],
    setEnableAnonymousUsers: [Function: setEnableAnonymousUsers]
  },
  schemaCache: {
    all: [Function: all],
    get: [Function: get],
    put: [Function: put],
    del: [Function: del],
    clear: [Function: clear]
  },
  _mount: 'http://127.0.0.1:3000/api',
  generateSessionExpiresAt: [Function: bound generateSessionExpiresAt],
  generateEmailVerifyTokenExpiresAt: [Function: bound generateEmailVerifyTokenExpiresAt],
  headers: {
    'user-agent': 'node-XMLHttpRequest, Parse/js1.11.1 (NodeJS 16.15.0)',
    accept: '*/*',
    'content-type': 'text/plain',
    host: '127.0.0.1:3000',
    'content-length': '146',
    connection: 'close'
  },
  ip: '::ffff:127.0.0.1'
}
express middleware___________________________________________________________________________________
undefined
parse middleware_________________________________________________________________________________
<ref *2> Config {
  applicationId: 'APP_ID',
  appId: 'APP_ID',
  appNAme: 'APP_NAME',
  javascriptKey: 'JAVASCRIPT_KEY',
.........................
```

dblythy commented:
Thanks for the explanation! I think this could be solved by #7869 - exposing the config via `config.get`. I think using express’ native syntax of `app.use` for middleware is a bit more intuitive than `ParseServer.middleware` and it’s easier to see the order of middleware.
